external/kafka/src/main/scala/org/apache/spark/rdd/kafka/KafkaRDD.scala

```diff
@@ -0,0 +1,199 @@
+    val batch: Array[KafkaRDDPartition],
+    messageHandler: MessageAndMetadata[K, V] => R
+  ) extends RDD[R](sc, Nil) with Logging {
+
+  override def getPartitions: Array[Partition] = batch.asInstanceOf[Array[Partition]]
+
+  override def getPreferredLocations(thePart: Partition): Seq[String] = {
+    val part = thePart.asInstanceOf[KafkaRDDPartition]
+    // TODO is additional hostname resolution necessary here
+    Seq(part.host)
+  }
+
+  override def compute(thePart: Partition, context: TaskContext): Iterator[R] = {
+    val part = thePart.asInstanceOf[KafkaRDDPartition]
+    if (part.fromOffset >= part.untilOffset) {
+      log.warn("Beginning offset is same or after ending offset " +
+        s"skipping ${part.topic} ${part.partition}")
+      Iterator.empty
+    } else {
+      new NextIterator[R] {
+        context.addTaskCompletionListener{ context => closeIfNeeded() }
+
+        log.info(s"Computing topic ${part.topic}, partition ${part.partition} " +
+          s"offsets ${part.fromOffset} -> ${part.untilOffset}")
+
+        val kc = new KafkaCluster(kafkaParams)
+        val keyDecoder = classTag[U].runtimeClass.getConstructor(classOf[VerifiableProperties])
+          .newInstance(kc.config.props)
+          .asInstanceOf[Decoder[K]]
+        val valueDecoder = classTag[T].runtimeClass.getConstructor(classOf[VerifiableProperties])
+          .newInstance(kc.config.props)
+          .asInstanceOf[Decoder[V]]
+        val consumer = connectLeader
+        var requestOffset = part.fromOffset
+        var iter: Iterator[MessageAndOffset] = null
+
+        // The idea is to use the provided preferred host, except on task retry atttempts,
+        // to minimize number of kafka metadata requests
+        private def connectLeader: SimpleConsumer = {
+          if (context.attemptNumber > 0) {
+            kc.connectLeader(part.topic, part.partition).fold(
+              errs => throw new Exception(
+                s"Couldn't connect to leader for topic ${part.topic} ${part.partition}: " +
+                  errs.mkString("\n")),
+              consumer => consumer
+            )
+          } else {
+            kc.connect(part.host, part.port)
+          }
+        }
+
+        private def handleErr(resp: FetchResponse) {
+          if (resp.hasError) {
+            val err = resp.errorCode(part.topic, part.partition)
+            if (err == ErrorMapping.LeaderNotAvailableCode ||
+              err == ErrorMapping.NotLeaderForPartitionCode) {
+              log.error(s"Lost leader for topic ${part.topic} partition ${part.partition}, " +
+                s" sleeping for ${kc.config.refreshLeaderBackoffMs}ms")
+              Thread.sleep(kc.config.refreshLeaderBackoffMs)
+            }
+            // Let normal rdd retry sort out reconnect attempts
+            throw ErrorMapping.exceptionFor(err)
+          }
+        }
+
+        override def close() = consumer.close()
+
+        override def getNext: R = {
```

Comment:
getNext should have parentheses (since it has side effect and it was defined with parentheses in NextIterator)